### ISSUE TITLE
async_mode=True giving wrong output

### ISSUE THREAD

**Issue by squarefk** [taichi-dev/taichi#2024]
```
import taichi as ti
import numpy as np

ti.init(arch=ti.cpu, async_mode=True)

x = ti.field(ti.f32, shape=(10,))
n = 10

@ti.kernel
def compute_mean_of_boundary_edges() -> ti.i32:
    total = 0.0
    for i in range(n):
        total += x[i] + x[i] * x[i]
    result = total / ti.cast(n, ti.i32)
    print("Mean of boundary edges:", result)
    return result


if __name__ == "__main__":
    x.from_numpy(np.array([0,1,2,3,4,5,6,7,8,9]))
    print(compute_mean_of_boundary_edges())
```

**Comment by xumingkuan** [comment]
Thanks for reporting this! AsyncTaichi does not handle global temps properly now.

A slightly smaller case:
```python
import taichi as ti
import numpy as np

ti.init(arch=ti.cpu, async_mode=True, print_ir=True)

x = ti.field(ti.f32, shape=(10,))
n = 10

@ti.kernel
def reduce() -> ti.f32:
    total = 0.0
    for i in range(n):
        total += x[i]
    print("Sum:", total)
    return total


if __name__ == "__main__":
    x.from_numpy(np.array([0,1,2,3,4,5,6,7,8,9]))
    print(reduce())
```
IR:
```
[I 11/07/20 00:33:38.479] [compile_to_offloads.cpp:taichi::lang::irpass::`anonymous-namespace'::make_pass_printer::<lambda_fe1d620add3df83d4ee306f9c0ab10ca>::operator ()@18] [reduce_c4_0] Simplified III:
kernel {
  $0 = offloaded  
  body {
    <*f32> $1 = global tmp var (offset = 0 B)
    <f32> $2 = const [0.0]
    <f32> $3 : global store [$1 <- $2]
  }
  $4 = offloaded range_for(0, 10) grid_dim=0 block_dim=32 
  body {
    <i32> $5 = loop $4 index 0
    <f32> $6 = global ptr [S2place<f32>], index [$5] activate=false
    <f32> $7 = global load $6
    <*f32> $8 = global tmp var (offset = 0 B)
    <f32> $9 = atomic add($8, $7)
  }
  $10 = offloaded  
  body {
    <*f32> $11 = global tmp var (offset = 0 B)
    <f32> $12 = global load $11
    print "Sum: ", $12, "\n"
    <f32> $14 : kernel return $12
  }
}
```

**Comment by k-ye** [comment]
Hi, could you try again? This specific problem should be fixed in #2061

**Comment by squarefk** [comment]
Yes, this code snippet works.
